tests/src/core/testqgslayoutmap.cpp

```diff
@@ -458,6 +459,39 @@ void TestQgsLayoutMap::dataDefinedStyles()
   QVERIFY( checker.testLayout( mReport, 0, 0 ) );
 }
 
+void TestQgsLayoutMap::dataDefinedCrs()
+{
+  QList<QgsMapLayer *> layers = QList<QgsMapLayer *>() << mRasterLayer << mPolysLayer << mPointsLayer << mLinesLayer;
+
+  QgsLayout l( QgsProject::instance() );
+  l.initializeDefaults();
+
+  QgsLayoutItemMap *map = new QgsLayoutItemMap( &l );
+  map->attemptMove( QgsLayoutPoint( 20, 20 ) );
+  map->attemptResize( QgsLayoutSize( 200, 100 ) );
+  map->setFrameEnabled( true );
+  map->setLayers( layers );
+  l.addLayoutItem( map );
+
+  QString def_crs = map->extent().asWktCoordinates();
+
+  //test epsg variable
+  map->dataDefinedProperties().setProperty( QgsLayoutObject::MapCrs, QgsProperty::fromExpression( QStringLiteral( "EPSG:2192" ) ) );
+  QString epsg = map->extent().asWktCoordinates();
+  QCOMPARE( epsg != def_crs, true );
+
+  //test proj string variable
+  map->dataDefinedProperties().setProperty( QgsLayoutObject::MapCrs, QgsProperty::fromExpression( QStringLiteral( "+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +wktext  +no_defs" ) ) );
+  QString projstr = map->extent().asWktCoordinates();
+  QCOMPARE( projstr != def_crs, true );
+
+  //test wkt variable
+  map->dataDefinedProperties().setProperty( QgsLayoutObject::MapCrs, QgsProperty::fromExpression( QStringLiteral( "PROJCS[\"ETRS89 / Belgian Lambert 2008\",GEOGCS[\"ETRS89\",DATUM[\"European Terrestrial Reference System 1989\",SPHEROID[\"GRS 1980\",6378137.0,298.257222101,AUTHORITY[\"EPSG\",\"7019\"]],TOWGS84[0.0,0.0,0.0,0.0,0.0,0.0,0.0],AUTHORITY[\"EPSG\",\"6258\"]],PRIMEM[\"Greenwich\",0.0,AUTHORITY[\"EPSG\",\"8901\"]],UNIT[\"degree\",0.017453292519943295],AXIS[\"Geodetic latitude\",NORTH],AXIS[\"Geodetic longitude\",EAST],AUTHORITY[\"EPSG\",\"4258\"]],PROJECTION[\"Lambert Conic Conformal (2SP)\",AUTHORITY[\"EPSG\",\"9802\"]],PARAMETER[\"central_meridian\",4.359215833333335],PARAMETER[\"latitude_of_origin\",50.79781500000001],PARAMETER[\"standard_parallel_1\",49.833333333333336],PARAMETER[\"false_easting\",649328.0],PARAMETER[\"false_northing\",665262.0],PARAMETER[\"standard_parallel_2\",51.16666666666667],UNIT[\"m\",1.0],AXIS[\"Easting\",EAST],AXIS[\"Northing\",NORTH],AUTHORITY[\"EPSG\",\"3812\"]]""" ) ) );
```

Comment:
I'd drop the WKT test -- creation of a CRS from WKT is well covered elsewhere, and this test will need a bunch of ifdefs to work around different results on different proj versions